lib/ticks-to-tree.js

```diff
@@ -99,71 +113,187 @@ function ticksToTree (ticks, options = {}) {
 
-    addToMergedTree(stack.map(({ S, name, value, top }) => ({ S, name, value, top })))
+    addToMergedTree(trees, stack.map(({ S, name, type, value, top }) => ({ S, name, type, value, top })))
     // mutate original (save another loop over stack + extra objects)
-    addToUnmergedTree(stack)
+    addToUnmergedTree(trees, stack)
   })
 
-  function addToMergedTree (stack) {
-    var lastFrame = null
-    stack.forEach((frame, ix) => {
-      // TODO need an elegant way to place inlined frames into *other* stacks
-      // where the function has disappeared – for now we just remove inlined
-      // stacks from merged graphs
-      if (frame.S > 2) return // skip inlined
-      if (ix > 0) lastFrame.children = lastFrame.children || []
-      const children = (ix === 0) ? merged.children : lastFrame.children
-      const child = children.find(({ name }) => name === frame.name)
-
-      if (child === undefined) children.push(frame)
-      else frame = child
-
-      if (ix === stack.length - 1) frame.top++
-      if (ix === 0) merged.value += 1
-      frame.value++
-
-      lastFrame = frame
+  let loadedSourceMaps = {}
+  if (sourceMaps) {
+    const mapper = async (jsLibrary) => {
+      loadedSourceMaps = loadSourceMap(jsLibrary, sourceMaps, loadedSourceMaps)
+      return loadedSourceMaps
+    }
+    await pMap(Object.keys(jsLibraries), mapper, { concurrency: 16 })
+  }
+  if (sourceMaps || relativePath) {
+    Object.keys(trees).map((tree) => {
+      trees[tree].children.map((frame, index) => {
+        renameFrame(options, frame, loadedSourceMaps)
+      })
     })
   }
+  Object.keys(loadedSourceMaps).forEach((file) => {
+    if (loadedSourceMaps[file] !== false) {
+      loadedSourceMaps[file].destroy()
+    }
+  })
+
+  return trees
+}
+
+async function loadSourceMap (fileName, sourceMaps, loadedSourceMaps) {
+  if (typeof loadedSourceMaps[fileName] === 'undefined') {
+    const sourceMapFiles = [
+      fileName + '.map' // dist.js -> dist.js.map
+    ]
+
+    if (typeof sourceMaps === 'object') {
+      if (sourceMaps[basename(fileName)]) {
+        sourceMapFiles.unshift(sourceMaps[basename(fileName)])
+      }
+      if (sourceMaps[fileName]) {
+        sourceMapFiles.unshift(sourceMaps[fileName])
+      }
+    }
+
+    if (existsSync(fileName)) {
+      // read last 2 lines to accomodate for a line-break
+      const lastLine = await readLastLines.read(fileName, 2)
+      const sourceMapMatch = lastLine.match(/\/\/[#@]\s+sourceMappingURL=([^\s]+)/i)
+      if (sourceMapMatch) {
+        sourceMapFiles.unshift(resolve(dirname(fileName), sourceMapMatch[1]))
+      }
+    }
+
+    for (let i = 0; i < sourceMapFiles.length; i++) {
+      if (sourceMapFiles[i]) {
+        // convert relative source map paths to absolute paths
+        if (!isAbsolute(sourceMapFiles[i])) {
+          sourceMapFiles[i] = resolve(dirname(fileName), sourceMapFiles[i])
+        }
+
+        if (existsSync(sourceMapFiles[i])) {
```

Comment:
Please use an async version of this.

Reply:
I was going to use fs.exists but the linter warned this is deprecated. Should I use fs.access instead?

Reply:
Using fs.access instead. :+1: